### DISCUSSION TITLE
Is it possible to add bare metal support to native-image?

### DISCUSSION THREAD

**Discussion by theoparis** [oracle/graal#8634] [Q&A]
I see that graal's native-image tool uses LLVM - would it be possible to support `-nostdlib` to enable java code to run on bare metal/embedded platforms?

> Finally, Native Image links compiled user code, parts of the Java runtime (for example, the garbage collector, threading support), and the results of code execution into the binary.

In this case it would mean having an option to customize or disable these if it is possible. My use case is not the best way to demonstrate this idea. I wanted to experiment with projects such as [brigadier](https://github.com/mojang/brigadier) on bare metal with custom I/O and memory allocation.

Additionally, it seems like graalvm only allows musl, glibc or bionic libc. Adding nostdlib would allow the user to custmize the libc.

See #393 - not sure why it was closed.

Attempting to add `-H:NativeLinkerOption="-nostdlib"`  gives errors related to JNI and malloc/free. I assume there is currently no way to disable JNI or jdk builtins such as canonicalize_md.c

**Comment by sgammon** [comment]
I believe [Cosmopolitan](https://github.com/jart/cosmopolitan) support might enable this functionality, which has been filed as #8350 / `GR-52220`.
